sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -167,6 +167,25 @@ abstract class BinaryExpression extends Expression with trees.BinaryNode[Express
 
   override def toString: String = s"($left $symbol $right)"
 
+  // Default behavior of evaluation according to the default nullability of BinaryExpression.
+  // If subclass of BinaryExpression override nullable, probably should also override this.
+  override def eval(input: InternalRow): Any = {
+    val evalE1 = left.eval(input)
+    if(evalE1 == null) {
+      null
+    } else {
+      val evalE2 = right.eval(input)
+      if (evalE2 == null) {
+        null
+      } else {
+        nullSafeEval(evalE1, evalE2)
+      }
+    }
+  }
+
+  protected def nullSafeEval(evalE1: Any, evalE2: Any): Any =
+    sys.error(s"BinaryExpressions must override either eval or nullSafeEval")
```

Comment:
Why not just leave it as unimplemented, so developer will notice that during coding, not runtime.